core/src/main/java/org/apache/calcite/sql/SqlCreateJoinIndex.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.sql;
+
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.util.ImmutableNullableList;
+
+import java.util.List;
+
+/**
+ * Parse tree for {@code CREATE JOIN INDEX} statement.
+ */
+public class SqlCreateJoinIndex extends SqlCreate {
+  private static final SqlOperator OPERATOR =
+      new SqlSpecialOperator("CREATE JOIN INDEX", SqlKind.CREATE_TABLE);
+
+  public final SqlIdentifier name;
+  public final List<SqlTableAttribute> tableAttributes;
+  public final SqlNodeList select;
+  public final SqlNode from;
+  public final SqlNode where;
+  public final SqlNodeList groupBy;
+  public final SqlNodeList orderBy;
+  public final List<SqlIndex> indices;
+
+  /** Creates a {@code SqlCreateJoinIndex}. */
+  public SqlCreateJoinIndex(SqlParserPos pos, SqlIdentifier name,
+      List<SqlTableAttribute> tableAttributes, SqlNodeList select, SqlNode from,
+      SqlNode where, SqlNodeList groupBy, SqlNodeList orderBy,
+      List<SqlIndex> indices) {
+    super(OPERATOR, pos, SqlCreateSpecifier.CREATE, false);
+    this.name = name;
+    this.tableAttributes = tableAttributes;
+    this.select = select;
```

Comment:
Sorry I didn't catch this sooner in the design doc, but rather than recreating all of the fields of the SELECT statement, can we leverage the existing SELECT parsing logic and nodes?

For example, similar to how OrderedQueryOrExpr returns a SqlNode object and sets that as the "source" in the SqlInsert() parsing method?

Reply:
Yeah I think that would be a lot simpler. Then I can remove a bunch of variables from SqlCreateJoinIndex and the copied methods from SqlSelect(). I'll work on refactoring this to use OrderedQueryOrExpr.